server/src/main/java/org/opensearch/cluster/routing/allocation/AwarenessReplicaBalance.java

```diff
@@ -0,0 +1,90 @@
+ *
+ * This is used in enforcing total copy of shard is a multiple of unique values of awareness attributes
+ * Helps in balancing shards across all awareness attributes and ensuring high availability of data.
+ */
+public class AwarenessReplicaBalance {
+    public static final Setting<Boolean> CLUSTER_ROUTING_ALLOCATION_AWARENESS_BALANCE_SETTING = Setting.boolSetting(
+        "cluster.routing.allocation.awareness.balance",
+        false,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
+    private volatile List<String> awarenessAttributes;
+
+    private volatile Map<String, List<String>> forcedAwarenessAttributes;
+
+    private volatile Boolean awarenessBalance;
+
+    public AwarenessReplicaBalance(Settings settings, ClusterSettings clusterSettings) {
+        this.awarenessAttributes = CLUSTER_ROUTING_ALLOCATION_AWARENESS_ATTRIBUTE_SETTING.get(settings);
+        clusterSettings.addSettingsUpdateConsumer(CLUSTER_ROUTING_ALLOCATION_AWARENESS_ATTRIBUTE_SETTING, this::setAwarenessAttributes);
+        setForcedAwarenessAttributes(CLUSTER_ROUTING_ALLOCATION_AWARENESS_FORCE_GROUP_SETTING.get(settings));
+        clusterSettings.addSettingsUpdateConsumer(
+            CLUSTER_ROUTING_ALLOCATION_AWARENESS_FORCE_GROUP_SETTING,
+            this::setForcedAwarenessAttributes
+        );
+        setAwarenessBalance(CLUSTER_ROUTING_ALLOCATION_AWARENESS_BALANCE_SETTING.get(settings));
+        clusterSettings.addSettingsUpdateConsumer(CLUSTER_ROUTING_ALLOCATION_AWARENESS_BALANCE_SETTING, this::setAwarenessBalance);
+
+    }
+
+    private void setAwarenessBalance(Boolean awarenessBalance) {
+        this.awarenessBalance = awarenessBalance;
+    }
+
+    private void setForcedAwarenessAttributes(Settings forceSettings) {
+        Map<String, List<String>> forcedAwarenessAttributes = new HashMap<>();
+        Map<String, Settings> forceGroups = forceSettings.getAsGroups();
+        for (Map.Entry<String, Settings> entry : forceGroups.entrySet()) {
+            List<String> aValues = entry.getValue().getAsList("values");
+            if (aValues.size() > 0) {
+                forcedAwarenessAttributes.put(entry.getKey(), aValues);
+            }
+        }
+        this.forcedAwarenessAttributes = forcedAwarenessAttributes;
+    }
+
+    private void setAwarenessAttributes(List<String> awarenessAttributes) {
+        this.awarenessAttributes = awarenessAttributes;
+    }
+
+    public int awarenessAttributes() {
+        int awarenessAttributes = 1;
+        if (this.awarenessBalance == false) {
+            return awarenessAttributes;
+        }
+        for (String awarenessAttribute : this.awarenessAttributes) {
+            if (forcedAwarenessAttributes.containsKey(awarenessAttribute)) {
+                awarenessAttributes = awarenessAttributes * forcedAwarenessAttributes.get(awarenessAttribute).size();
+            }
+        }
+        return awarenessAttributes;
+    }
+
```

Comment:
What happens when we have multiple awareness attributes like zone/rack/switch with a configuration like
N0 -> Zone-A, Rack-A
N1  -> Zone-A, Rack-B
N2 -> Zone-B, Rack-A
N3  -> Zone-B, Rack-B


as compared to

N0 -> Zone-A, Rack-M
N1  -> Zone-B, Rack-N
N2 -> Zone-C, Rack-O
N3  -> Zone-D, Rack-P

Do we need to multiply or should we simply ensure that all copies are distributed across their respective attribute values. I would recommend we treat all attributes separately

Reply:
This makes sense @Bukhtawar  . I have corrected the logic now to take the longest varying attribute only 

N0 -> Zone-A, Rack-M
N1 -> Zone-B, Rack-N
N2 -> Zone-C, Rack-O
N3 -> Zone-D, Rack-P

In the above case now we would recommend total 4 copies .